Util Slave and Better Reprovision Process

jacderida opened the pull request:
Hi Stephen,

This PR has 3 different stages. The first is to verify the refactored Terraform code is OK, the 2nd is to test the util slave and the 3rd is to test that the environment remains accessible during the reprovisioning process. I was toying with the idea of doing the S3 backend for the state files, but I think that should probably be a separate thing.

In terms of issues:

* This closes #96
* This closes #107
* This closes #108

# First Stage

The first stage will verify that the refactored Terraform code should behave as expected.

## Instructions

* Clone a copy of the `maidsafe/safe-build-infrastructure` repo and leave it on the `master` branch
* Clone a copy of the `jacderida/safe-build-infrastructure` fork and switch to the `util_slave` branch
* From the `master` branch of the `maidsafe/safe-build-infrastructure` do `cd terraform/staging && terraform init && terraform apply`
* Delete any pre-existing `.terraform` directory or `.tfstate` files in the `jacderida/safe-build-infrastructure` fork if you already had it cloned
* Copy the .tfstate file resulting from the staging environment creation to the `jacderida/safe-build-infrastructure` repo and put it under `terraform/staging`
* From the `util_slave` branch of the `jacderida/safe-build-infrastructure` do `cd terraform/staging && terraform init && terraform plan` (note the `plan`, *not* `apply`)

## Acceptance Criteria

The output from the `terraform plan` should report 13 resources to add:
```
+ aws_instance.util_slave
    <snipped>

+ aws_key_pair.util_slave
    <snipped>

+ aws_security_group.util_slaves
    <snipped>

+ aws_security_group_rule.haproxy_ingress_ssh_from_util_slave
    <snipped>

+ aws_security_group_rule.jenkins_master_egress_ssh_to_util_slaves
    <snipped>

+ aws_security_group_rule.jenkins_master_ingress_ssh_from_util_slave
    <snipped>

+ aws_security_group_rule.util_slaves_egress_http
    <snipped>

+ aws_security_group_rule.util_slaves_egress_https
    <snipped>

+ aws_security_group_rule.util_slaves_egress_ssh_to_haproxy
    <snipped>

+ aws_security_group_rule.util_slaves_egress_ssh_to_jenkins_master
    <snipped>

+ aws_security_group_rule.util_slaves_egress_winrm_to_windows_slaves
    <snipped>

+ aws_security_group_rule.util_slaves_ingress_ssh_from_jenkins_master
    <snipped>

+ aws_security_group_rule.windows_slaves_ingress_winrm_from_util_slave
    <snipped>
Plan: 13 to add, 0 to change, 0 to destroy.
```

# Second Stage

This stage will create the staging environment and use the util slave to run a `docker_build` job to construct a container.

## Instructions

* Destroy the staging environment created in the first stage
* Re-create a new environment using the usual instructions
* Login to Jenkins and populate the jobs with the job seed file from the `jenkins-jobs` repo
* Put the `docker_slave` key on the Bastion under `~/.ssh/docker_slave` and run `chmod 0400 ~/.ssh/docker_slave`
* Run the `docker_build-safe_client_libs_build_container` job (not sure of the time: could take around 20 minutes to run through)
* While the job is running, get the private IP address of the util slave and attempt to SSH from the Bastion with `ssh -i ~/.ssh/docker_slave jenkins@<util slave private ip>`

## Acceptance Criteria

* The environment should be created with no errors
* The environment should be provisioned with no errors
* No manual setup should be required for dynamic slaves (private keys are now inserted automatically)
* The `docker_build` job should run on the util slave
* The `docker_build` job should complete successfully
* You should *not* be able to SSH from the Bastion to the util slave

# Third Stage

## Instructions

This is to make sure the environment remains online during a reprovision, which is the process we're now using for updating the prod environment. Previously it was going offline.

* Go back to the Bastion host and edit the Makefile to put some pause statements in the `provision-staging-jenkins-aws-reprovision`:
  - After `run_ansible_against_haproxy.sh` add `read -p 'Press any key to continue'`
  - After `run_ansible_against_jenkins_master.sh` add `read -p 'Press any key to continue'`
  - After `-u ansible ansible/haproxy-ssl-config.yml` add `read -p 'Press any key to continue'`
* Run `make provision-staging-jenkins-aws-reprovision`
* When the process pauses refresh `jenkins-staging.maidsafe.net` in the browser

## Acceptance Criteria

* The environment should reprovision successfully
* At the first pause Jenkins should still be reachable in the browser
* At the second pause Jenkins should still be reachable in the browser
* At the third pause Jenkins should still be reachable in the browser
* The Jenkins GUI should *not* indicate a restart

jacderida committed:
feat/env: initial attempt at util slave

This is a slave that's going to be used for deployments and such things.
The difficulty with it is going to be the security group configuration.
This is the first initial attempt, which is opening a connection between
the Bastion and the util slave (which will be removed) and trying to
establish internet connectivity.

jacderida committed:
fix/env: move from directory to flat files

It turns out Terraform wasn't picking up the files in the util_slave
directory, or at least it wasn't trying to apply any of the resources. I
think they count as a module, and I don't have time to refactor for that
at the moment.

jacderida committed:
feat/env: rules for util slave connectivity

This enables the following connectivity for the util slave:

* SSH connectivity from util_slaves -> linux_slaves
* SSH connectivity from util_slaves -> jenkins_master
* SSH connectivity from util_slaves -> haproxy
* WINRM connectivity from util_slaves -> windows_slaves

All the security groups and rules are also refactored to
use their own files. It was becoming far too difficult to navigate the
single file. Will also move other stuff to their own files as part of
this branch.

jacderida committed:
refactor/env: move docker_slave_ssh_public_key

This moves this variable from the `all` group to the `linux_slaves` and
`osx_slaves` groups. This variable also really needs to have a more
generic name, like `jenkins_slave_ssh_public_key`, because it won't be
specific to the Docker slave any more.

The reason why it's being moved is because the `util_slaves` group needs
to provide a different value for this key.

jacderida committed:
feat/box: packer template for util slave

Provides a template and a target for running the template.

In the provisioners, you can see some strange things going on
with the installation of Python. For some reason Ubuntu 18.04 doesn't
ship with an installation of Python, but for some unknown reason,
sometimes when you try to install it here, it complains that there's no
package available for it. This is the reason for having the sed command
that edits the repository sources. Sadly though, this doesn't appear to
fix it all the time. You need to just keep running the target until it
works.

If the problem continues to persist, we can create a base AMI that just
has Python installed.

jacderida committed:
Merge branch 'master' into util_slave

jacderida committed:
refactor/box: use script to install python

jacderida committed:
fix/slave: provide private key

This gets the key into the correct format needed by CASC. You need to
use the indenting function to push the content of the variable over to
the right, otherwise CASC won't accept it.

jacderida committed:
refactor/slave: new 'cloud' for util slave

Although the EC2 plugin allows you to define any number of 'templates'
under each cloud entry, it seems it only allows you to define one
private key for all the templates. For this reason, each slave has been
added under a separate cloud entry.

jacderida committed:
fix/slave: correct keys and group names

jacderida committed:
feat/env: connectivity for util slave

This opens connectivity between the jenkins_master and the util_slave
security groups and private subnets. This is necessary for the master
instance to SSH to the instance on the util slave group and private
subnet.

jacderida committed:
fix/box: install java on slave

Without this, Jenkins itself will attempt to install Java when it
connects to the machine, but the jenkins user doesn't have root access,
so the process fails.

jacderida committed:
fix/box: use vault pass for ansible run

jacderida committed:
feat/box: new slave ami ID

jacderida committed:
refactor/role: docker-slave-images

Changes the role from just pulling in docker images to also logging in
to Dockerhub so that the machine would be able to push an image.

There may also need to be some other things added at some point in the
future, and it feels better to put this functionality in here rather
than the jenkins-slave role, as you wouldn't necessarily want to login
or even have Docker installed on a slave.

jacderida committed:
feat/env: new util slave AMI ID

This has the Jenkins user logged into Dockerhub, which should allow the
user to be able to push a new container.

jacderida committed:
refactor/env: split qa environment into files

jacderida committed:
refactor/env: split prod environment into files

jacderida committed:
fix/env: jenkins inaccessible during reprovision

There was an issue with Jenkins going offline during the reprovisioning
process. This was due to HAProxy not forwarding the request, due to a
bad configuration.

There are 2 different configurations for HAProxy: one in this role and
one in the haproxy-ssl-config role. If you look at the first commit for
the haproxy-ssl-config role, it discusses why, so I won't repeat that
here. During the reprovisioning process, the reason why HAProxy was
going offline is because the first config file is overwriting the
second one, which it doesn't like because the first config file doesn't
have SSL enabled.

The issue has been resolved by providing a 'run_mode' variable that
indicates whether we're doing an initial provision or a reprovision. The
first config file will only be copied during the initial run.

jacderida committed:
Merge remote-tracking branch 'upstream/master' into util_slave

jacderida committed:
feat/role: add safe cli image to docker slave

jacderida committed:
refactor/env: remove temporary connectivity

Remove the temporary connectivity between the Bastion and the util
slave.

jacderida committed:
fix/env: correct names for security groups

jacderida committed:
fix/env: remove static util slave

This was added when the initial design was to have this as a static
slave, but later on it turned out to make more sense to have this
machine run as a dynamic slave, so this static instance is no longer
needed.

S-Coyle committed:
fix/env: correct environment names in security groups

S-Coyle committed:
refactor/env: delete unnecessary reboot_slaves method

The reboot_slaves method was duplicated by the reboot_all_instances.sh
script in the initial provision, and no reboots are necessary in the
reprovision, so we are able to delete this method.

S-Coyle committed:
docs/readme: update readme to reflect most recent changes

S-Coyle committed:
refactor/env: update utils_slave machine size and macos executors

Update utils_slave machine size to `T2Medium`.
Update number of executors on macOS to now be 2.

S-Coyle commented:
Stage 1:
- [x] Acceptance criteria met with the same 13 resources listed as in the description above.

Stage 2:
- [x] The environment should be created with no errors.
- [x] The environment should be provisioned with no errors.
- [x] No manual setup should be required for dynamic slaves (private keys are now inserted automatically).
- [x] The docker_build job should run on the util slave.
- [x] The docker_build job should complete successfully.
- [x] You should not be able to SSH from the Bastion to the util slave.

Stage 3:
- [x] As agreed with Chris, I tested this differently to the steps above - I reprovisioned while the docker build from stage 2 was underway. Reprovisioning was successful and the docker build progress was available, and progressing, throughout.

S-Coyle reviewed (dismissed):
👍

jacderida committed:
fix/script: remove method call

This method was removed a few commits ago but the reference to it wasn't
taken out.

jacderida commented:
I've pushed a tiny little commit with a fix, so I'm just going to merge that without an approval because I want to deploy this tonight.
